lib/rubocop/cop/rspec/rails/minitest_assertions.rb

```diff
@@ -229,6 +219,48 @@ def assertion
               'be_empty'
             end
           end
+
+          MSG = 'Use `%<prefer>s`.'
+
+          # TODO: replace with `BasicAssertion.subclasses` in Ruby 3.1+
+          ASSERTION_MATCHERS = constants(false).filter_map do |c|
+            const = const_get(c)
+
+            const if const.is_a?(Class) && const.superclass == BasicAssertion
+          end
+
+          RESTRICT_ON_SEND = ASSERTION_MATCHERS.flat_map { |m| m::MATCHERS }
+
+          ASSERTION_MATCHERS.each do |m|
+            name = m.name.split('::').last
+
+            def_node_matcher "minitest_#{name}".to_sym, m::NODE_MATCHER_PATTERN
```

Comment:
If you add `extend NodePattern::Macros` to `BasicAssertion`, it should be possible to add the node matchers in the `BasicAssertion` subclasses themselves. I’m not 100% sure if it works, but we’d get rid of this loop defining the node patterns, and `#on_send` would also become a lot simpler, e.g.
```rb
def on_send(node)
  ASSERTION_MATCHERS.each do |matcher|
    assertion = matcher.new(node).match
    return if assertion.nil?

    on_assertion(node, assertion)
  end
end
```

Reply:
ah cool - @pirj made a similar comment over in #1791 but I'd not explored further since I thought it'd require a more complicated change to the constructor, but maybe not.

I'll give that a try but it if turns out too complicated I would prefer to tackle it in a follow-up PR if that's ok

Reply:
hmm yeah this is turning ugly a bit too ugly for my taste: since `BasicAssertion` is handling a lot of the instance variables, we effectively want an artificial second constructor, and the node matcher method takes a block which makes it messier to do a return (unless I've missed something).

Would you mind sticking with this current layout and I can explore this further as a follow up?

Reply:
Sure, we can stick with the current layout.

I tried finding a suitable implementation, and it’s tricky. I ended up with something like this in the subclasses:
```rb
            def initialize(node)
              pattern(node) do |actual, failure_message|
                super(nil, actual, failure_message.first)
              end
            end
```
The “sometimes calling `super`” smells a bit, and I had to add this method to the `BasicAssertion`:
```
            def match?
              !!@expected || !!@actual
            end
```
to be called instead of `next if assertion.nil?`.

Reply:
Yeah it's annoying how ugly it gets and I'm pretty sure it's all because of `node_matcher` not being on the class... 😅

I've got an idea or two on how to approach it though which I'll explore after this is landed.